ethereum/core/src/main/java/org/hyperledger/besu/ethereum/transaction/TransactionSimulator.java

```diff
@@ -380,4 +374,87 @@ public Optional<Boolean> doesAddressExist(
 
     return Optional.of(worldState.get(address) != null);
   }
+
+  @FunctionalInterface
+  public interface SimulationGasLimitCalculator {
+    long calculate(final long txGasLimit, final long blockGasLimit);
+
+    static long bounded(final long txGasLimit, final long blockGasLimit, final long rpcGasCap) {
+      final long txGasCap = txGasLimit >= 0 ? txGasLimit : blockGasLimit;
+
+      if (rpcGasCap > 0) {
+        final long simulationGasLimit = Math.min(txGasCap, rpcGasCap);
+        LOG.trace(
+            "Gas limit capped at {} for transaction simulation, tx gas limit={}, block gas limit={}, RPC gas cap={}",
+            simulationGasLimit,
+            txGasLimit,
+            blockGasLimit,
+            rpcGasCap);
+        return simulationGasLimit;
+      }
+      return txGasLimit;
+    }
+
+    static long unbounded(final long txGasLimit, final long blockGasLimit, final long rpcGasCap) {
+
+      if (rpcGasCap > 0) {
+        LOG.trace("Returning RPC gas cap {} for transaction simulation", rpcGasCap);
+
+        return rpcGasCap;
+      }
+
+      final long simulationGasLimit = txGasLimit >= 0 ? txGasLimit : blockGasLimit;
+
+      LOG.trace(
+          "Gas limit capped at {} for transaction simulation, tx gas limit={}, block gas limit={}",
+          simulationGasLimit,
+          txGasLimit,
+          blockGasLimit);
+      return simulationGasLimit;
+    }
+  }
+
+  public enum RpcGasCapMode {
+    BOUNDED,
+    UNBOUNDED
+  }
+
+  public static final class Builder {
+    private final Blockchain blockchain;
+    private final WorldStateArchive worldStateArchive;
+    private final ProtocolSchedule protocolSchedule;
+    private long rpcGasCap = 0;
+    private RpcGasCapMode rpcGasCapMode = RpcGasCapMode.BOUNDED;
+
+    public Builder(
+        final Blockchain blockchain,
+        final WorldStateArchive worldStateArchive,
+        final ProtocolSchedule protocolSchedule) {
+      this.blockchain = blockchain;
+      this.worldStateArchive = worldStateArchive;
+      this.protocolSchedule = protocolSchedule;
+    }
+
+    public Builder rpcGasCap(final long rpcGasCap, final RpcGasCapMode mode) {
+      this.rpcGasCap = rpcGasCap;
+      this.rpcGasCapMode = mode;
+      return this;
+    }
+
+    public TransactionSimulator build() {
+      final SimulationGasLimitCalculator simulationGasLimitCalculator;
+      if (rpcGasCap > 0 && rpcGasCapMode.equals(RpcGasCapMode.UNBOUNDED)) {
```

Comment:
a `long` + enum encodes same amount of information as `Long` or Optional<Long> null or `Optional.empty()` may mean no bound. Introduction of `RpcGasCapMode` is verbose, but maybe it makes it more obvious

Reply:
However possibility to set it to (0, BOUNDED) makes it errorprone. Optional or a nullable Long is less errorprone from this perspective